### ISSUE TITLE
Issue with vkCmdPipelineBarrier and depth buffers.

### ISSUE THREAD

**Issue by Onhi** [GPUOpen-Drivers/AMD-Gfx-Drivers#8]
Something is up with vkCmdPipelineBarrier... It's messing up the content of depth buffers.

GPU: RX 7900XTX
Driver Version: 31.0.24033.1003
OS: Windows 11 Pro (10.0.22000 Build 22000)

Here's an exemple:

```
srcStageMask         VK_PIPELINE_STAGE_2_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_2_LATE_FRAGMENT_TESTS_BIT
srcAccessMask        VK_ACCESS_2_MEMORY_READ_BIT | VK_ACCESS_2_MEMORY_WRITE_BIT
dstStageMask         VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT
dstAccessMask        VK_ACCESS_2_MEMORY_READ_BIT
oldLayout            VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL
newLayout            VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL
```

![before](https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/assets/19506696/3da0d7dd-ee70-4bd0-b215-7c391292b996)

![after](https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/assets/19506696/81d9869a-230f-4e5d-b4a6-d9f300567d38)

I know this looks like a problem with the application but I almost certain it is not. Vulkan validation layers are not showing any warning/errors, RenderDoc can capture and replay the issue... And the same code is running flawlessly on team green hardware.

I can help finding the issue providing time, renderdoc captures, vulkan api dumps, code (possible but tricky), etc...

Thanks!

**Comment by owenzhangzhengzhong** [comment]
Hi @Onhi,
Can you provide compilable source code that reproduces this problem?
Anything else you can provide would also be helpful.
Thanks,
Owen

**Comment by Onhi** [comment]
I'll get to reducing the engine to a simple case demonstrating the issue. In the meantime, here's a link to a renderdoc capture
(compatible with an RX 7900XTX) and a reduced API dump of the issue. In the following, the resource "camera_depth" is the one showcasing the issue.
https://drive.google.com/drive/folders/13M_uPHq1WdBmEf0FVYVS0QKMV86O7pkb?usp=drive_link

**Comment by Onhi** [comment]
I got a version of the code ready, where should I put it?

**Comment by jinjianrong** [comment]
@Onhi You may create a repository in your github space https://github.com/Onhi?tab=repositories to put your code

**Comment by Onhi** [comment]
I cant put the code on github. I've create a rar file in the google drive folder I linked just above for you to retrieve the code.

**Comment by owenzhangzhengzhong** [comment]
Thanks @Onhi, once I get access to the folder I will create internal ticket for this issue.

**Comment by Onhi** [comment]
@owenzhangzhengzhong you should be able to access the folder now.

**Comment by owenzhangzhengzhong** [comment]
Hey @Onhi, 
Able to access the folder, and was able to compile and see the issue in the editor on my local system. I've created internal ticket to track this issue, and we'll look into it. In the meantime, is it possible for you to simplify the testcase for this issue? There's a lot of source to go through, if you can reduce it as much as possible to just rendering with the vkCmdPipelineBarrier that's causing the corruption that would make it much easier for us to investigate.
Thanks,
Owen

**Comment by Onhi** [comment]
Cool, thanks Owen, I'll do my best to reduce the code to a minimum. Might take some time but I'll get to it during the weekend.

**Comment by Onhi** [comment]
Hi, I'm having an hard time reducing the engine more than what I have provided while still providing a solid repro case (we need to move a camera to change the range of captured information and or resize images... so its not trivial to isolate).

Doing more tests, I see irregular behavior around depth/stencil images & layout transition barriers. On some captures, its the depth that is blocky and bugged (like in the rdc capture I provided) other times it's the stencil aspect of the image that is destroyed.

Full disclaimer, in my effort to reduce the size of the code, I noticed an issue with one of the barrier pair in the code I provided. Around the OIT pass, I wrongfully removed the pair of barriers that where converting from shader to depth stencil and vice versa. Still, fixing it had no impact on the issue.

I understand that this is a foreign code base so I would like to offer help debugging the issue if needed. (via video conf. or otherwise).

**Comment by owenzhangzhengzhong** [comment]
Hi @Onhi,
Cursory look with synchronization validation enabled shows some errors:
SYNC-HAZARD-READ-AFTER-WRITE
SYNC-HAZARD-WRITE-AFTER-WRITE
Associated with camera_depth and camera_image_view, can you look at how you're setting up the pipeline barriers associated with those command submissions. See attached full log of validation errors:
[Editor.txt](https://github.com/user-attachments/files/15779619/Editor.txt)
Meanwhile we'll look further as well,
Owen

**Comment by Onhi** [comment]
Fixed the validation errors using store op none (instead of store) on the read only usage of the camera_depth buffer in the OIT pass. 

Sadly it didnt fix the main issue.

**Comment by owenzhangzhengzhong** [comment]
Did you fix all the validation issues? Can you post that code as well? With details on which lines you updated?

**Comment by Onhi** [comment]
Hi, I've uploaded a v2 of the reduced engine with fixes for validation errors and details of lines changed.

**Comment by Onhi** [comment]
I've updated the version again with more fixes (V3). This version can go through queue synchronization validatation & synchronization2 layer without error/warnings.
- I shuffled a ferw descriptor updates and added the update after bind flag on the descriptor sets,
- Double & tripple checked all barriers... they should all be setup correctly now.
- Fixed the presentation engine semaphores' wait/signal pipeline stage flags.

The problem with depth is still unresolved. :(

**Comment by owenzhangzhengzhong** [comment]
Hi @Onhi,
Brief update, was able to not see the issue with triggering CmdBarrier call after every command driver side:
Before:
![image](https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/assets/98582551/99cfeb77-b7e4-4c07-85ef-7969ddc2a5a2)

After:
![image](https://github.com/GPUOpen-Drivers/AMD-Gfx-Drivers/assets/98582551/33c57927-a800-40ce-ac17-30c8f8551aff)
 (This is with resize the windows for a bit)

After narrowing it down a bit it seems related to missing barriers after commands associated with direct dispatch and transfer copies involving an image. Still narrowing it down further.

**Comment by Onhi** [comment]
Wow! Very nice progress! Thanks for the update.

**Comment by owenzhangzhengzhong** [comment]
Hi @Onhi,

After narrowing it down further I see the issue resolved by adding this barrier:

```
VkMemoryBarrier mem_barrier{};
mem_barrier.sType         = VK_STRUCTURE_TYPE_MEMORY_BARRIER;
mem_barrier.srcAccessMask = VK_ACCESS_2_COLOR_ATTACHMENT_READ_BIT;
mem_barrier.dstAccessMask = VK_ACCESS_2_NONE;
 
vkCmdPipelineBarrier(command_buffer,
                     VK_PIPELINE_STAGE_NONE,
                     VK_PIPELINE_STAGE_2_TOP_OF_PIPE_BIT_KHR,
                     0,
                     1,
                     &mem_barrier,
                     0,
                     nullptr,
                     0,
                     nullptr);
```
In the file: AMDReproEngine\Oasis\code\Engine\Vulkan\Vulkan.Interface.h
In the following functions after the Vulkan API calls:
void copyImage(...
After: vkCmdCopyImage2(command_buffer, &copy_image_info);

void copyBufferToImage(...
After: vkCmdCopyBufferToImage2(command_buffer, &copy_buffer_to_image_info);

void copyImageToBuffer(...
After: vkCmdCopyImageToBuffer2(command_buffer, &copy_image_to_buffer_info);

void blitImage(...
After: vkCmdBlitImage2(command_buffer, &blit_image_info);

void dispatch(...
After: vkCmdDispatch(command_buffer, p_group_count_x, p_group_count_y, p_group_count_z);

I see issue mostly resolved after adding the barrier in copyImage and dispatch. I suspect further optimizations can be made.

**Comment by Onhi** [comment]
Hi Owen, so what is the rationale for this fix?

Not sure I understand the meaning of the barrier and why its needed in all those places.

**Comment by Onhi** [comment]
In plain english this barrier mean:

> All color attachment reads from the 1st set of commands (regardless which stage they are in) must be finished before the 2nd set of commands start executing.
> Effectively speaking, since no stage is considered within the 1st set of commands, then the GPU is free to "go ahead" and execute the 2nd set of commands even though the 1st set of commands may have not completed.

I understand that it may produce an hw barrier in the driver that fixes the issue but I think its still indicative of a driver issue.

**Comment by owenzhangzhengzhong** [comment]
Hi @Onhi 
I'm not familiar with the specifics of how you're handing rendering, but the resolution via missing barriers indicates that there's an issue with application synchronization causing corruptions.
It looks like dispatch and copyImage needs to be finished before execution of the graphics pipeline involving framebuffer reads. Top of pipe is probably too restrictive and can be further optimized.
Vulkan is pretty permissive in memory access and allows a lot of leeway in synchronization. Different drivers could handle synchronization differently on the backend. So, drivers of other hardware could be less permissive in handling synchronization. So it could be that a driver implementation/hardware hides a missing application barrier, while another driver has optimized unnecessary synchronization or cache operations, thus exposing the missing barrier in the application and the resulting corruption.
There’s also a lot of per application barrier optimizations which you can see some of the code here:
https://github.com/GPUOpen-Drivers/xgl/tree/dev/icd/api/appopt
